File: 20有效的括号.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s)%2==1: return False

        stack = []

        pairs = {
            "}" : "{",
            "]" : "[",
            ")" : "("
        }

        
        for char in s:
            if char in pairs:
                if not stack or pairs[char] != stack[-1]:
                    return False
                stack.pop()
            else:
                stack.append(char)

        
        return not stack

File: 20有效的括号.swift
class Solution {
    func isValid(_ s: String) -> Bool {
        if s.count % 2 == 1 { return false }
        let dic = ["{":"}",
                "(":")",
                "[":"]"]
        let chars = Array(s)
        var stack = [String]()
        for char in chars {
            if dic[String(char)] != nil {
                stack.append(String(char))
            } else {
                if stack.isEmpty || dic[stack[stack.count - 1]] != String(char) {
                    return false
                }
                stack.popLast()
            }
        }
        return stack.isEmpty
    }
}
